src/parse_util.h

```diff
@@ -0,0 +1,130 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ *
+ */
+
+#pragma once
+
+#include <cstdlib>
+#include <limits>
+#include <string>
+#include <tuple>
+#include <status.h>
+
+namespace details {
+
+template <typename>
+struct ParseIntFunc;
+
+template <>
+struct ParseIntFunc<short> { // NOLINT
+  constexpr static const auto value = std::strtol;
+};
+
+template <>
+struct ParseIntFunc<int> {
+  constexpr static const auto value = std::strtol;
+};
+
+template <>
+struct ParseIntFunc<long> { // NOLINT
+  constexpr static const auto value = std::strtol;
+};
+
+template <>
+struct ParseIntFunc<long long> { // NOLINT
+  constexpr static const auto value = std::strtoll;
+};
+
+template <>
+struct ParseIntFunc<unsigned short> { // NOLINT
+  constexpr static const auto value = std::strtoul;
+};
+
+template <>
+struct ParseIntFunc<unsigned> {
+  constexpr static const auto value = std::strtoul;
+};
+
+template <>
+struct ParseIntFunc<unsigned long> { // NOLINT
+  constexpr static const auto value = std::strtoul;
+};
+
+template <>
+struct ParseIntFunc<unsigned long long> { // NOLINT
+  constexpr static const auto value = std::strtoull;
+};
+
+}  // namespace details
+
+template <typename T>
+using ParseResultAndPos = std::tuple<T, const char *>;
+
+template <typename T = long long> // NOLINT
+StatusOr<ParseResultAndPos<T>> TryParseInt(const char *v, int base = 0) {
+  char *end;
+
+  errno = 0;
+  auto res = details::ParseIntFunc<T>::value(v, &end, base);
+
+  if (v == end) {
+    return {Status::NotOK, "TryParseInt: not started as an integer"};
+  }
+
+  if (errno) {
+    return {Status::NotOK, std::string {"TryParseInt: "} + std::strerror(errno)};
+  }
+
+  if (!std::is_same<T, decltype(res)>::value &&
```

Comment:
This is unnecessary, if the parsed result is out or range, error will be set as ERANGE.

Reply:
Hi, this is actually necessary. If `T` is `int` or `short`, `strtol` will be used, so `decltype(res)` is `long` (there is no `strtoi` or `strtos`). Actually the implementation in libstdc++ do the totally same thing as here. 

https://code.woboq.org/gcc/libstdc++-v3/include/ext/string_conversions.h.html#73

Reply:
What if `res` and `T` are signed and unsigned? Guess more careful type check is needed here.

Reply:
Hi @fuzhe1989 , it will not happen, you could check here:
https://github.com/apache/incubator-kvrocks/pull/787/files#diff-01299a1bd002d7db4c7c4e6c059d1a6a97ac2abea9042acafc5473e474f87d85R34-R72

Reply:
got it

Reply:
@PragmaTwice You are right, I have ignored the return type of ParseIntFunc<T>::value, I miss understand it and think the return type is T.

I think the value of `end` should be checked, if end != v + strlen(v), this means we meet something like "123abc". In this situation should we return an error ?

Reply:
This is the difference between `TryParseInt` and `ParseInt`. `ParseInt` check it and return error while `end != input + size`, but `TryParseInt` just return the `end` pointer, to cover more situation.